parsec/basicclient.go

```diff
@@ -12,35 +12,63 @@ import (
 	"github.com/parallaxsecond/parsec-client-go/parsec/algorithm"
 )
 
+// providerSelector provides functionality to select core and crypto providers - supplied as
+// implementation of operations.ProviderSelector to the underlying operations client
+type providerSelector struct {
+	implicitProvider ProviderID
+}
+
+// GetProvider gets the correct provider to use for an opcode.  Implements operations.ProviderSelector
+func (ps *providerSelector) GetProvider(opcode requests.OpCode) requests.ProviderID {
+	switch opcode { //nolint:exhaustive // we handle the exhaustive case correctly
+	case requests.OpPing, requests.OpListProviders, requests.OpListOpcodes, requests.OpListKeys, requests.OpListAuthenticators:
+		return requests.ProviderCore
+	default:
+		return requests.ProviderID(ps.implicitProvider)
+	}
+}
+
+// defaultProviderSelector returns providerSelector with a defult implicit provider
+func defaultProviderSelector() *providerSelector {
+	return &providerSelector{implicitProvider: ProviderMBed}
+}
+
 // BasicClient is a Parsec client representing a connection and set of API implementations
 type BasicClient struct {
-	opclient *operations.Client
+	opclient         *operations.Client
+	providerSelector *providerSelector
 }
 
 // InitClient initializes a Parsec client
 func InitClient() (*BasicClient, error) {
-	opclient, err := operations.InitClient()
+	selector := defaultProviderSelector()
+	opclient, err := operations.InitClient(selector)
```

Comment:
Following my other open comment, `defaultProviderSelector` or similar could instead return a `ProviderID` which it passes to the `InitClient` constructor. It would be something similar as well for the authentication type! I think that avoids going back and forth between the layers.